ui-framework-documentation.md

```diff
@@ -113,100 +177,42 @@ sub run {
 }
 ```
 
+Or if we are using [external test data](https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/declarative-schedule-doc.md#test_data) (recommended)
+
+```perl
+sub run {
+    my $test_data = get_test_suite_data()
+    my $partitioner = $testapi::distri->get_partitioner();
+    
+    foreach my $partition (@{test_data->{partitions}) { $partitioner->partition_disk($partition) }
+}
+ ```
 #### Access to os-autoinst testapi from Test Module
 
 Test Module is not allowed to use os-autoinst testapi functions
 directly. It should use methods, provided by [Controller](#controller)
 layer instead. This allows to hide the details of the UI structure and
-operate with the business logic the system provides.
+operate with the [business
+logic](https://en.wikipedia.org/wiki/Business_logic) the system provides.
+
+This being said, for logging purposes, "diag", "record_info" or "take_screenshot", "record_soft_failure" can be used
+in the test module when it cannot be done at the page level.
 
 #### Access to other framework layers from Test Module
 
 Test Module is able to interact only with the Controller layer.
 
 ### Controller
 
-Controller is a layer that provides methods to interact with the system
-under test in business terms.
-
-#### Test Data usage in Controller
-
-Do not define any test data in [Controller](#controller) layer as it
-could make test maintenance more complicated. Use the test data passed
-from the [Test Module](#test-module) layer instead.
-
-Example:
-
-* If need to use conditions: 
-
-    ```perl
-    sub create_encrypted_partition {
-        my ($self, $is_lvm) = @_;
-        if ($is_lvm) {
-            $self->get_partitioning_scheme_page()->select_lvm_checkbox();
-        }
-    ...
-    }
-    ```
-* To make an action depending on a value:
-
-  
-    ```perl
-    sub create_filesystem {
-        my ($self, $filesystem) = @_;
-        $self->get_filesystem_options_page()->select_filesystem($filesystem);
-    ...
-    }
-    ```
-    
-#### Access to os-autoinst testapi from Controller
-
-Ideally, it should not use the os-autoinst testapi directly. 
-
-* Do not use testapi methods that communicates with the SUT (e.g.
-  `send_keys`, `assert_screen`). Wrap them into [Page](#page) methods
-  with the meaningful names instead.
-
-* Using `get_var` to change the flow of a test or get a data for the
-  test should be avoided as much as possible (e.g. to decide whether
-  check or uncheck checkbox, use method parameters instead and pass the
-  data from [Test Module](#test-module).
-
-**NOTE:** Since os-autoinst API does not have separation between methods
```

Comment:
I looked in all our controllers and found only one example of using testapi, on an old controller, some record_info I think. So I think we can safely remove that part.

Reply:
added